trunk-recorder/main.cc

```diff
@@ -954,6 +954,75 @@ void handle_call(TrunkMessage message, System *sys) {
 
 }
 
+void clear_stale_moto_patches(System *sys){
+  std::map<int,std::time_t> patch_map;
+  std::vector<std::map<int,std::time_t>> current_patches = sys->get_talkgroup_patches();
+  std::vector<std::map<int,std::time_t>> updated_patches;
+  //for each patch map in the vector:
+  BOOST_FOREACH (auto& patch, current_patches) {
+    //for each TGID/element in the patch map:
+    patch_map.clear();
+    BOOST_FOREACH(auto& patch_element, patch){
+      if (std::time(nullptr) - patch_element.second < 3){  //3 second hard coded timeout for now
+        //only add back into our active patch map if the timestamp associated with this TGID is less than three seconds old
+        patch_map[patch_element.first] = patch_element.second; 
+      }
+    }
+    if(!patch_map.empty()){
+      updated_patches.push_back(patch_map);
+    }
+  }
+  BOOST_LOG_TRIVIAL(info) << "Found " << updated_patches.size() << " active Moto patches";
+  BOOST_FOREACH (auto& patch, updated_patches) {
+    string printstring;
+    BOOST_FOREACH(auto& patch_element, patch){
+      printstring+=" ";
+      printstring+= std::to_string(patch_element.first);
+    }
+    BOOST_LOG_TRIVIAL(info) << "Active Moto Patch of TGIDs" << printstring;
+  }
+  sys->set_talkgroup_patches(updated_patches);
+}
+
+void update_moto_patches(TrunkMessage message, System *sys){
+  std::map<int,std::time_t> patch_map;
```

Comment:
Yes, I didn't want to assume that the entire patch goes up and down at the same time.  That may be the case, but due to my ignorance I thought it would be safer to have a "last heard" timestamp mapped to each TG in the patch so that if individual TGs are added or dropped while a single super group is still active, it would handle those gracefully.  The map also easily deals with something I see often on my local system - when a TG of only two patches is active, it still populates all four fields (sg, ga1, ga2, ga3)...the map structure handles the duplication (only one element gets put in the map for each TGID...if there are duplicates, it just updates the existing TGID entry). 

I'm not sure if there is value in identifying the supergroup or not.  While a typical subscriber radio needs this to know which TGID to look for coming over the air, trunk-recorder's operation is a bit different.  Even if we don't track which TG is the supergroup in the patch list, it can still be implied as being the TG that's actually used over the air.  So when we get to the point of having a call active, the TGID of the Call is the supergroup, by definition.

Reply:
ok - that makes sense. I was thinking of just using a Struct that would have have: a timeout for the Patch, the super group id, and a vector of the talkgroups. This would only work if new patches are created when a TG is removed.

Let's give the map approach a try. It will give us a better sense of a patches lifecycle.

Reply:
I refactored this quite a bit in my latest commit, I think it's a bit simpler to follow now.  Still not modifying the map in place during iteration, but quite a bit cleaner I think.